Add integration tests for aws lightsail

prasadkatti opened the pull request:
##### SUMMARY
Integration test suite for lightsail module

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lightsail

##### ADDITIONAL INFORMATION
This patch updates the aws IAM policy (compute-policy).

The lightsail module requires a keypair to create instances. I have created a keypair named `lightsail` in my account that I used while writing these tests. 
NOTE: A keypair will need to be created in the test account for the test suite to work.

prasadkatti committed:
Add integration tests for aws lightsail

prasadkatti commented:
@jillr - I have added this test suite for lightsail module as promised on IRC. :)
@mattclay - This test suite updates the compute policy for AWS.

s-hertel commented on hacking/aws_config/testing_policies/compute-policy.json:
```diff
@@ -245,6 +245,19 @@
             ]
+        },
+        {
+            "Sid": "AllowLightsail",
```
If you want to also add a PR to this repo and file https://github.com/mattclay/aws-terminator/blob/master/aws/policy/compute.yaml with these, it's where we deploy the permissions for CI. The hacking/ directory policies make running the tests in contributors' own accounts more convenient but the permissions aren't all strict enough to actually use for CI.

prasadkatti replied:
Hi @s-hertel. I took a crack at it, but I am not very sure about the change. Please let me know if this works -
https://github.com/mattclay/aws-terminator/pull/68

jillr commented on test/integration/targets/lightsail/tasks/main.yml:
```diff
@@ -0,0 +1,136 @@
+          aws_secret_key: "{{ aws_secret_key }}"
+          security_token: "{{ security_token }}"
+          region: "{{ aws_region }}"
+      no_log: yes
```
As an FYI, we're trying to move more towards using module_defaults for credentials in new tests.  This works just fine for this PR though.
https://docs.ansible.com/ansible/devel/dev_guide/platforms/aws_guidelines.html#aws-credentials-for-integration-tests

prasadkatti replied:
I can make that change to use module_defaults.

jillr commented on test/integration/targets/lightsail/tasks/main.yml:
```diff
@@ -0,0 +1,136 @@
+
+    - assert:
+        that:
+          - result.changed == False
```
It would be nice to check that the state is running or pending here as well (like you do in the next start task further down).  You could even add a `wait: yes` to the previous task to validate that the waiter function does the right thing, and then only assert on `running`.

prasadkatti replied:
okay, I will add the `wait: yes`. 👍

prasadkatti replied:
I don't think the `wait` option is working as one would expect. The `wait` option defaults to `yes` for this module. I ran a task to create instance with `wait: yes`, but the module returned with `state=pending`.
Looking at the [code](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/lightsail.py#L316), wait is being used to let the instance come out of a transient state (pending, stopping).  The `create_instance` function is not looking at `wait` at all.

prasadkatti replied:
So should I add a failing test or should I add it and comment it out for now?

prasadkatti replied:
The description for the `wait` option in the module documentation is also incorrect. It says -

> Wait for the instance to be in state 'running' before returning. If wait is "no" an ip_address may not be returned

But that is not the case. I ran a task to create instance and it returned this as the state -
`        "state": {
            "code": 0,
            "name": "pending"
        },
`

jillr replied:
Ah apologies, I naively trusted the module docs.  For now let's just assert that the instance is in ['pending', 'running'].  I've opened #63869 about the lack of wait support for create actions.

prasadkatti replied:
I have added the test case to check the instance state is in ['pending', 'running'].

prasadkatti committed:
lightsail - use module_defaults instead of aws_connection_info

prasadkatti committed:
lightsail tests - assert instance state on create

prasadkatti commented:
Hi @jillr and @s-hertel. I believe I have added patches to take care of review comments. I have also added a PR to update the compute policy in aws-terminator. This is ready_for_review.

prasadkatti commented:
@jillr / @s-hertel - Hello 👋. Are either of you planning on merging this in this week? I have updated the aws-terminator and I believe this is good to go now.

jillr commented on test/integration/targets/lightsail/tasks/main.yml:
```diff
@@ -0,0 +1,123 @@
+      security_token: '{{ security_token | default(omit) }}'
+      region: '{{ aws_region | default(omit) }}'
+
+- name: Integration test for lightsail module
```
If you look at the details of the CI failure, you have a YAML syntax failure. This line is not necessary.
```suggestion
```

jillr reviewed (commented):
Hi @prasadkatti, I generally go through the terminator queue on Thursdays, usually every other week but I'll be going through a few tomorrow.

prasadkatti committed:
Fix yaml syntax error

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

prasadkatti commented:
My test setup is still a bit hacky :( . I need to work on that. Is this
something that sanity checks would have caught?

On Wed, Nov 6, 2019 at 9:44 AM Jill R <notifications@github.com> wrote:

> *@jillr* commented on this pull request.
>
> Hi @prasadkatti <https://github.com/prasadkatti>, I generally go through
> the terminator queue on Thursdays, usually every other week but I'll be
> going through a few tomorrow.
> ------------------------------
>
> In test/integration/targets/lightsail/tasks/main.yml
> <https://github.com/ansible/ansible/pull/63770#discussion_r343234481>:
>
> > @@ -0,0 +1,123 @@
>
> +---
>
> +
>
> +- module_defaults:
>
> +    group/aws:
>
> +      aws_access_key: '{{ aws_access_key | default(omit) }}'
>
> +      aws_secret_key: '{{ aws_secret_key | default(omit) }}'
>
> +      security_token: '{{ security_token | default(omit) }}'
>
> +      region: '{{ aws_region | default(omit) }}'
>
> +
>
> +- name: Integration test for lightsail module
>
>
> If you look at the details of the CI failure, you have a YAML syntax
> failure. This line is not necessary.
> ⬇️ Suggested change
>
> -- name: Integration test for lightsail module
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/63770?email_source=notifications&email_token=ABYBIYNAV22G6NDYCQCAZODQSL7BDA5CNFSM4JDIXV22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCKRDQAQ#pullrequestreview-312621058>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABYBIYOW2KNVAGWB4QUBJ6DQSL7BDANCNFSM4JDIXV2Q>
> .
>

jillr commented:
No because it's valid yaml, just not valid ansible. If you have an AWS account you can use, you can run the integration tests locally by following: https://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html
Note that the config file locaton has moved to https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/config/cloud-config-aws.ini.template though.
Sanity tests can also be run locally like: `ansible-test sanity test/integration/targets/lightsail/`
Or just a specific test as: `ansible-test sanity --test yamllint  test/integration/targets/lightsail/`

prasadkatti commented:
Hi @jillr. I have updated my local test setup using the blog post you shared above. I was able to verify the latest commit using that setup. So that's good.  Although, we still need a keypair that we can use when creating an instance. The lightsail module does not support creating a keypair. So I am guessing that will needed to be created using some different means. In my local test setup, I have a keypair called 'lightsail' created that I use for all test runs.

Question for you - Is aws-terminator deployed on a certain schedule? Or is that done on an as-needed basis? Will I need to manually kick off ci for this PR once the new policies are deployed?

prasadkatti commented:
The only thing missing for this is a keypair to use when creating instances and then this should be good to go.

mattclay commented:
@prasadkatti Can you update the test to create the key pair it requires? In the absence of a feature complete Ansible module for creating a key pair, a minimal implementation can be added to the `library` directory of the test. That keeps the test self-contained and will avoid the need to pre-generate a key in the account used for running tests.

prasadkatti committed:
[lightsail] create keypair as part of the testsuite

prasadkatti commented:
@mattclay - I have updated the test to create the key pair as you suggested. I added a module called `lightsail_keypair` in the library directory of the test. It is a minimal module that only allows you to create a keypair. If the keypair exists already, the module exits with `changed=False`. The module has been added to the `aws` group of module_defaults.

The keypair itself has been named `lightsail_ci`. The same keypair will be used across all the test runs for this testsuite. This requires the CI account to allow `lightsail:CreateKeyPair` action. I have added the statement in the relevant policy in https://github.com/mattclay/aws-terminator/pull/76.

prasadkatti committed:
Fix lightsail actions in compute-policy

mattclay commented:
@prasadkatti To avoid leaving resources behind after the tests finish, update the test to use a unique name for the keypair and remove it at the end of the test using a handler.

Also update your PR https://github.com/mattclay/aws-terminator/pull/76 to include permissions for removal, and add a class to remove keypairs. That way the terminator can clean up after aborted tests.

prasadkatti commented:
Sounds good. That was the direction I was originally heading towards, but then I thought that it would be easier to make all test runs use the same keypair.

I will make the updates and push out a patch ASAP.

prasadkatti committed:
Add ability to delete keypair in lightsail_keypair

prasadkatti commented:
ready_for_review
@mattclay

jillr commented:
This lgtm, I've got the terminator change planned to test tomorrow but at a quick glance it looks good as well.

jillr reviewed (approved):
terminator changes have been deployed to prod, thanks for your work on this @prasadkatti

jillr commented:
rebuild_merge

prasadkatti commented:
Thanks for all the reviews and help with this PR @jillr @mattclay @s-hertel. This was a fun learning experience. I would definitely like to do this with more modules that are missing tests.
